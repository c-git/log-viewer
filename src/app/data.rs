use std::collections::BTreeMap;

use anyhow::Context;

#[derive(serde::Deserialize, serde::Serialize, Default, Debug)]
pub struct Data {
    pub selected_row: Option<usize>,
    rows: Vec<LogRow>,
}

#[derive(serde::Deserialize, serde::Serialize, Default, Debug)]
pub struct LogRow {
    time: Option<String>,
    request_id: Option<String>,
    #[serde(rename = "otel.name")]
    otel_name: Option<String>,
    msg: Option<String>,

    #[serde(flatten)]
    pub extra: BTreeMap<String, serde_json::Value>,
}

impl LogRow {
    const TEXT_FOR_EMPTY: &'static str = "[ --- ]";

    pub(crate) fn time(&self) -> &str {
        if let Some(val) = &self.time {
            val
        } else {
            Self::TEXT_FOR_EMPTY
        }
    }

    pub(crate) fn request_id(&self) -> &str {
        if let Some(val) = &self.request_id {
            val
        } else {
            Self::TEXT_FOR_EMPTY
        }
    }

    pub(crate) fn otel_name(&self) -> &str {
        if let Some(val) = &self.otel_name {
            val
        } else {
            Self::TEXT_FOR_EMPTY
        }
    }

    pub(crate) fn msg(&self) -> &str {
        if let Some(val) = &self.msg {
            val
        } else {
            Self::TEXT_FOR_EMPTY
        }
    }
}

impl Data {
    pub fn rows(&self) -> &[LogRow] {
        &self.rows
    }
}

impl TryFrom<&str> for Data {
    type Error = anyhow::Error;

    fn try_from(value: &str) -> Result<Self, Self::Error> {
        let mut result = Data::default();
        for (i, line) in value.lines().enumerate() {
            result.rows.push(
                serde_json::from_str(line)
                    .with_context(|| format!("failed to parse line {}", i + 1))?,
            );
        }
        Ok(result)
    }
}

#[cfg(test)]
mod tests {
    use insta::glob;
    use rstest::{fixture, rstest};

    use super::*;

    const PATH_PROJECT_ROOT: &str = "../../";
    const PATH_TEST_SAMPLES: &str = "tests/sample_logs/*.*";

    #[fixture]
    pub(crate) fn insta_settings() -> insta::Settings {
        let mut result = insta::Settings::clone_current();
        let cwd = std::env::current_dir().expect("failed to get cwd");
        let path = cwd.join("tests").join("snapshots");
        result.set_snapshot_path(path);
        result
    }

    #[rstest]
    fn deserialize_rows_from_string(insta_settings: insta::Settings) {
        glob!(PATH_PROJECT_ROOT, PATH_TEST_SAMPLES, |path| {
            let input = std::fs::read_to_string(path).unwrap();
            let data = Data::try_from(&input[..]).unwrap();
            insta_settings.bind(|| insta::assert_ron_snapshot!(data));
        });
    }
}
