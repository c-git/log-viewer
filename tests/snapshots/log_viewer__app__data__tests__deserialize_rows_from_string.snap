---
source: src/app/data.rs
expression: data
---
Data(
  selected_row: None,
  rows: [
    {
      "time": Some("2024-02-10T03:13:04.191299188Z"),
      "request_id": None,
      "otel.name": None,
      "msg": Some("Server address is: 127.0.0.1:8000"),
      "file": "crates/server/src/startup.rs",
      "hostname": "my_pc",
      "level": 30,
      "line": 34,
      "name": "my_server",
      "pid": 50401,
      "target": "my_server::startup",
      "v": 0,
    },
    {
      "time": Some("2024-02-10T03:13:04.191610465Z"),
      "request_id": None,
      "otel.name": None,
      "msg": Some("starting 8 workers"),
      "file": "/home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/actix-server-2.3.0/src/builder.rs",
      "hostname": "my_pc",
      "level": 30,
      "line": 240,
      "name": "my_server",
      "pid": 50401,
      "target": "actix_server::builder",
      "v": 0,
    },
  ],
)
